src/app/clusters/dishwasher-alarm-server/dishwasher-alarm-server.cpp

```diff
@@ -0,0 +1,397 @@
+    {
+        ChipLogProgress(Zcl, "Dishwasher Alarm: ERR: Mask is not supported");
+        return EMBER_ZCL_STATUS_FAILURE;
+    }
+
+    EmberAfStatus status = EMBER_ZCL_STATUS_SUCCESS;
+    if ((status = Attributes::Mask::Set(endpoint, mask)) != EMBER_ZCL_STATUS_SUCCESS)
+    {
+        ChipLogProgress(Zcl, "Dishwasher Alarm: ERR: writing  mask, err:0x%x", status);
+        return status;
+    }
+
+    // Whenever there is change in Mask, State should change accordingly.
+    BitMask<AlarmMap> state;
+    status = GetStateValue(endpoint, &state);
+    if (status != EMBER_ZCL_STATUS_SUCCESS)
+    {
+        return status;
+    }
+
+    if (!mask.HasAll(state))
+    {
+        state  = mask & state;
+        status = SetStateValue(endpoint, state, true);
+    }
+    return status;
+}
+
+EmberAfStatus DishwasherAlarmServer::SetLatchValue(EndpointId endpoint, const BitMask<AlarmMap> latch)
+{
+    if (!HasResetFeature(endpoint))
+    {
+        ChipLogProgress(Zcl, "Dishwasher Alarm feature: Unsupport Latch attribute");
+        return EMBER_ZCL_STATUS_UNSUPPORTED_ATTRIBUTE;
+    }
+
+    BitMask<AlarmMap> supported;
+    if (GetSupportedValue(endpoint, &supported) || !supported.HasAll(latch))
+    {
+        ChipLogProgress(Zcl, "Dishwasher Alarm: ERR: Latch is not supported");
+        return EMBER_ZCL_STATUS_FAILURE;
+    }
+
+    EmberAfStatus status = Attributes::Latch::Set(endpoint, latch);
+    if (status != EMBER_ZCL_STATUS_SUCCESS)
+    {
+        ChipLogProgress(Zcl, "Dishwasher Alarm: ERR: writing  latch, err:0x%x", status);
+        return status;
+    }
+
+    return status;
+}
+
+EmberAfStatus DishwasherAlarmServer::SetStateValue(EndpointId endpoint, const BitMask<AlarmMap> newState, bool ignoreLatchState)
+{
+    BitMask<AlarmMap> supported;
+    BitMask<AlarmMap> finalNewState;
+    finalNewState.Set(newState);
+
+    if (GetSupportedValue(endpoint, &supported) || !supported.HasAll(finalNewState))
+    {
+        ChipLogProgress(Zcl, "Dishwasher Alarm: ERR: Alarm is not supported");
+        return EMBER_ZCL_STATUS_FAILURE;
+    }
+
+    BitMask<AlarmMap> mask;
+    if (GetMaskValue(endpoint, &mask) || !mask.HasAll(finalNewState))
+    {
+        ChipLogProgress(Zcl, "Dishwasher Alarm: ERR: Alarm is suppressed");
+        return EMBER_ZCL_STATUS_FAILURE;
+    }
+
+    EmberAfStatus status = EMBER_ZCL_STATUS_SUCCESS;
+    BitMask<AlarmMap> currentState;
+    status = Attributes::State::Get(endpoint, &currentState);
+    if (status != EMBER_ZCL_STATUS_SUCCESS)
+    {
+        ChipLogProgress(Zcl, "Dishwasher Alarm: ERR: reading  state, err:0x%x", status);
+        return status;
+    }
+
+    BitMask<AlarmMap> latch;
+    if (!ignoreLatchState && (GetLatchValue(endpoint, &latch) == EMBER_ZCL_STATUS_SUCCESS))
+    {
+        // only reset bits which can be reset based on latch value
+        BitMask<AlarmMap> finalState(latch & currentState, finalNewState);
+        finalNewState = finalState;
```

Comment:
Isn't this backwards?  This only allows resetting bits that are _set_ in Latch, but the intended thing (and the documentation for the function) says that it should allow bits that are _clear_ in Latch.  So I think this should be:
```suggestion
        // Only reset bits which are not latched.
        finalNewState.Clear(latch);
```

Reply:
No, the logic (L&C|N) does all of the logic needed to determine the state of all of the bits in the attribute.  I didn't know about the Clear method until after I figured out that logic, so the assignment is correct.  Of course, I have no way to test, so...

Reply:
Oh, I completely missed the `| finalNewState` bit there!  How about writing it out a bit more clearly as:
```
    auto bitsToKeep = latch & currentState;
    finalNewState.Set(bitsToKeep);
```
or something like that?

Reply:
Can we please add the Or operator to the BitFlags template?

I'll update the code per what you are saying...

Reply:
Code updated per comments and discussion on Slack with @bzbarsky-apple.